### PR TITLE
[SPARK-32861][SQL] GenerateExec should require column ordering

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?
This PR updates the `RemoveRedundantProjects` rule to make `GenerateExec` require column ordering. 

### Why are the changes needed?
`GenerateExec` was originally considered as a node that does not require column ordering. However, `GenerateExec` binds its input rows directly with its `requiredChildOutput` without using the child's output schema. 
In `doExecute()`:
```scala
val proj = UnsafeProjection.create(output, output)
```
In `doConsume()`:
```scala
val values = if (requiredChildOutput.nonEmpty) {
  input
} else {
  Seq.empty
}
```
In this case, changing input column ordering will result in `GenerateExec` binding the wrong schema to the input columns. For example, if we do not require child columns to be ordered, the `requiredChildOutput` [a, b, c] will directly bind to the schema of the input columns [c, b, a], which is incorrect:
```
GenerateExec explode(array(a, b, c)), [a, b, c], false, [d]
  HashAggregate(keys=[a, b, c], functions=[], output=[c, b, a])
    ...
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Unit test

### CODE REVIEW

**Comment by viirya** [comment]
The PR description is not clear enough. Can you revise it? It should be nicer to describe which part in `GenerateExec` has the issue. An example is better.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala [hunk 1]
```diff
@@ -18,17 +18,21 @@
...
-import org.apache.spark.sql.test.{SharedSparkSession, SQLTestUtils}
+import org.apache.spark.sql.test.SharedSparkSession
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.util.Utils
 
-class RemoveRedundantProjectsSuite extends QueryTest with SharedSparkSession with SQLTestUtils {
+abstract class RemoveRedundantProjectsSuiteBase
+  extends QueryTest
+    with SharedSparkSession
+    with AdaptiveSparkPlanHelper {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala:30]: Is this PR related to AQE? (the PR description does say nothing about it though)

**Comment by allisonwang-db** [comment]
@viirya Good point! I've revised the description and please let me know if it's unclear.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala (hunk 1 above)
Comment (allisonwang-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala:30]: This rule was also added in AQE's [query stage prep rules](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala#L88) so the test should also make sure the rule works when AQE is enabled.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala [hunk 2]
```diff
@@ -115,9 +119,40 @@ class RemoveRedundantProjectsSuite extends QueryTest with SharedSparkSession wit
...
+          g.copy(requiredChildOutput = requiredChildOutput.reverse,
+            child = ProjectExec(requiredChildOutput.reverse, child))
+      }
+
+      // Re-apply remove redundant project rule.
+      val rule = RemoveRedundantProjects(spark.sessionState.conf)
+      val newExecutedPlan = rule.apply(newPlan)
+      // The manually added ProjectExec node shouldn't be removed.
+      assert(collectWithSubqueries(newExecutedPlan) {
+        case p: ProjectExec => p }.size == numProjects + 1)
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala:148]: The style looks weird.

```scala
assert(collectWithSubqueries(newExecutedPlan) {
  case p: ProjectExec => p
}.size == numProjects + 1)

```

**Review by viirya — APPROVED** [review]
Looks good except for one comment about code style.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala (hunk 2 above)
Comment (allisonwang-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala:148]: Updated.

**Review by maropu — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by dongjoon-hyun** [comment]
Hi, all.
SPARK-32861 seems to have affected version 3.0.1.
Do we need to backport this to branch-3.0?

**Comment by cloud-fan** [comment]
if it's a bug in 3.0, yea let's backport

**Comment by dongjoon-hyun** [comment]
Thank you so much, @cloud-fan .
